File: app/src/main/java/com/hiddenramblings/tagmo/browser/WebsiteFragment.java
package com.hiddenramblings.tagmo.browser;

import android.app.Dialog;
import android.app.ProgressDialog;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.util.Base64;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.JavascriptInterface;
import android.webkit.WebResourceRequest;
import android.webkit.WebResourceResponse;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.widget.EditText;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.Fragment;
import androidx.webkit.ServiceWorkerClientCompat;
import androidx.webkit.ServiceWorkerControllerCompat;
import androidx.webkit.WebViewAssetLoader;
import androidx.webkit.WebViewClientCompat;
import androidx.webkit.WebViewFeature;

import com.hiddenramblings.tagmo.NFCIntent;
import com.hiddenramblings.tagmo.R;
import com.hiddenramblings.tagmo.TagMo;
import com.hiddenramblings.tagmo.amiibo.AmiiboManager;
import com.hiddenramblings.tagmo.eightbit.io.Debug;
import com.hiddenramblings.tagmo.eightbit.os.Storage;
import com.hiddenramblings.tagmo.nfctech.TagUtils;

import org.json.JSONException;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.text.ParseException;
import java.util.regex.Pattern;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

public class WebsiteFragment extends Fragment {

    private WebView mWebView;
    private ProgressDialog dialog;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        return inflater.inflate(R.layout.activity_webview, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        mWebView = view.findViewById(R.id.webview_content);
        WebSettings webViewSettings = mWebView.getSettings();

        mWebView.setScrollbarFadingEnabled(true);
        webViewSettings.setLoadWithOverviewMode(true);
        webViewSettings.setUseWideViewPort(true);
        webViewSettings.setAllowFileAccess(true);
        webViewSettings.setAllowContentAccess(false);
        webViewSettings.setJavaScriptEnabled(true);
        webViewSettings.setDomStorageEnabled(true);
        webViewSettings.setCacheMode(WebSettings.LOAD_NO_CACHE);
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.KITKAT) {
            webViewSettings.setPluginState(WebSettings.PluginState.ON);
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            final WebViewAssetLoader assetLoader =
                    new WebViewAssetLoader.Builder().addPathHandler("/assets/",
                            new WebViewAssetLoader.AssetsPathHandler(requireContext())).build();

            mWebView.setWebViewClient(new WebViewClientCompat() {
                @Override
                public WebResourceResponse shouldInterceptRequest(
                        WebView view, WebResourceRequest request) {
                    return assetLoader.shouldInterceptRequest(request.getUrl());
                }
            });
            if (WebViewFeature.isFeatureSupported(WebViewFeature.SERVICE_WORKER_BASIC_USAGE)) {
                ServiceWorkerControllerCompat.getInstance().setServiceWorkerClient(
                        new ServiceWorkerClientCompat() {
                            @Nullable
                            @Override
                            public WebResourceResponse shouldInterceptRequest(
                                    @NonNull WebResourceRequest request) {
                                return assetLoader.shouldInterceptRequest(request.getUrl());
                            }
                        });
            }
        }
        webViewSettings.setAllowFileAccessFromFileURLs(
                Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP);
        webViewSettings.setAllowUniversalAccessFromFileURLs(
                Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP);

        JavaScriptInterface download = new JavaScriptInterface();
        mWebView.addJavascriptInterface(download, "Android");
        mWebView.setDownloadListener((url, userAgent, contentDisposition, mimeType, contentLength) -> {
            if (url.startsWith("blob") || url.startsWith("data")) {
                Log.d("DATA", url);
                mWebView.loadUrl(download.getBase64StringFromBlob(url, mimeType));
            }
        });

        loadWebsite(null);
    }

    public void loadWebsite(String address) {
        if (null != mWebView) {
            if (null == address) address = NFCIntent.SITE_GITLAB_README;
            WebSettings webViewSettings = mWebView.getSettings();
            if (address.startsWith(NFCIntent.SITE_GITLAB_README)) {
                webViewSettings.setUserAgentString(webViewSettings.getUserAgentString().replaceAll(
                        "(?i)" + Pattern.quote("android"), "TagMo"));
            } else {
                webViewSettings.setBuiltInZoomControls(true);
                webViewSettings.setSupportZoom(true);
                webViewSettings.setDefaultZoom(WebSettings.ZoomDensity.FAR);
            }
            mWebView.loadUrl(address);
        } else {
            final String delayedUrl = address;
            new Handler(Looper.getMainLooper()).postDelayed((Runnable) () ->
                    loadWebsite(delayedUrl), TagMo.uiDelay);
        }
    }

    private class UnZip implements Runnable {
        File archive;
        File outputDir;

        UnZip(File ziparchive, File directory) {
            this.archive = ziparchive;
            this.outputDir = directory;
        }

        private final Handler handler = new Handler(Looper.getMainLooper());
        @SuppressWarnings("ResultOfMethodCallIgnored")
        @Override
        public void run() {
            try {
                ZipInputStream zipIn = new ZipInputStream(new FileInputStream(this.archive));
                ZipEntry entry;
                while (null != (entry = zipIn.getNextEntry())) {
                    ZipEntry finalEntry = entry;
                    handler.post(() -> dialog.setMessage(
                            getString(R.string.unzip_item, finalEntry.getName())));
                    if (finalEntry.isDirectory()) {
                        File dir = new File(outputDir, finalEntry.getName()
                                .replace("/", ""));
                        if (!dir.exists() && !dir.mkdirs())
                            throw new RuntimeException(
                                    getString(R.string.mkdir_failed, dir.getName())
                            );
                    } else {
                        FileOutputStream fileOut = new FileOutputStream(
                                new File(outputDir, finalEntry.getName()));
                        byte[] buffer = new byte[8192];
                        int len;
                        while ((len = zipIn.read(buffer)) != -1)
                            fileOut.write(buffer, 0, len);
                        fileOut.close();
                        zipIn.closeEntry();
                    }
                }
                zipIn.close();
            } catch (IOException e) {
                Debug.Log(e);
            } finally {
                dialog.dismiss();
                this.archive.delete();
            }
        }
    }

    private void unzipFile(File zipFile) {
        dialog = ProgressDialog.show(requireContext(), "", "", true);
        new Thread(new UnZip(zipFile, Storage.getDownloadDir(
                "TagMo", "Downloads"))).start();
    }

    private void saveBinFile(byte[] tagData, String name) {
        try {
            File filePath = new File(Storage.getDownloadDir("TagMo", "Downloads"),
                    name + ".bin");
            FileOutputStream os = new FileOutputStream(filePath, false);
            os.write(tagData);
            os.flush();
        } catch (IOException e) {
            Debug.Log(e);
        }
    }

    private void setBinName(String base64File, String mimeType) {
        byte[] tagData = Base64.decode(base64File.replaceFirst(
                "^data:" + mimeType + ";base64,", ""), 0);
        View view = getLayoutInflater().inflate(R.layout.dialog_backup, null);
        AlertDialog.Builder dialog = new AlertDialog.Builder(requireContext());
        final EditText input = view.findViewById(R.id.backup_entry);
        try {
            AmiiboManager amiiboManager = AmiiboManager
                    .getAmiiboManager(requireContext().getApplicationContext());
            input.setText(TagUtils.decipherFilename(amiiboManager, tagData, true));
        } catch (IOException | JSONException | ParseException e) {
            Debug.Log(e);
        }
        Dialog backupDialog = dialog.setView(view).create();
        view.findViewById(R.id.save_backup).setOnClickListener(v -> {
            saveBinFile(tagData, input.getText().toString());
            backupDialog.dismiss();
        });
        view.findViewById(R.id.cancel_backup).setOnClickListener(v -> backupDialog.dismiss());
        backupDialog.show();
    }

    private class JavaScriptInterface {
        @SuppressWarnings("unused")
        @JavascriptInterface
        public void getBase64FromBlobData(String base64Data) throws IOException {
            convertBase64StringSave(base64Data);
        }
        public String getBase64StringFromBlob(String blobUrl, String mimeType) {
            if (blobUrl.startsWith("blob") || blobUrl.startsWith("data")) {
                return "javascript: var xhr = new XMLHttpRequest();" +
                        "xhr.open('GET', '" + blobUrl + "', true);" +
                        "xhr.setRequestHeader('Content-type','" + mimeType + "');" +
                        "xhr.responseType = 'blob';" +
                        "xhr.onload = function(e) {" +
                        "  if (this.status == 200) {" +
                        "    var blobFile = this.response;" +
                        "    var reader = new FileReader();" +
                        "    reader.readAsDataURL(blobFile);" +
                        "    reader.onloadend = function() {" +
                        "      base64data = reader.result;" +
                        "      Android.getBase64FromBlobData(base64data);" +
                        "    }" +
                        "  }" +
                        "};" +
                        "xhr.send();";
            }
            return "javascript: console.log('Not a valid blob URL');";
        }
        private void convertBase64StringSave(String base64File) throws IOException {
            String zipType = getString(R.string.mimetype_zip);
            if (base64File.contains("data:" + zipType + ";")) {
                File filePath = new File(Storage.getDownloadDir("TagMo"), "download.zip");
                FileOutputStream os = new FileOutputStream(filePath, false);
                os.write(Base64.decode(base64File.replaceFirst(
                        "^data:" + zipType + ";base64,", ""), 0));
                os.flush();
                unzipFile(filePath);
            } else {
                String[] binTypes = getResources().getStringArray(R.array.mimetype_bin);
                for (String binType : binTypes) {
                    if (base64File.contains("data:" + binType + ";")) {
                        setBinName(base64File, binType);
                        break;
                    }
                }
            }
        }
    }
}

File: app/src/main/java/com/hiddenramblings/tagmo/browser/WebsiteFragment.kt
package com.hiddenramblings.tagmo.browser

import android.annotation.SuppressLint
import android.app.Dialog
import android.app.ProgressDialog
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Base64
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.webkit.*
import android.widget.EditText
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.Fragment
import androidx.webkit.*
import androidx.webkit.WebViewAssetLoader.AssetsPathHandler
import com.hiddenramblings.tagmo.NFCIntent
import com.hiddenramblings.tagmo.R
import com.hiddenramblings.tagmo.TagMo
import com.hiddenramblings.tagmo.amiibo.AmiiboManager.Companion.getAmiiboManager
import com.hiddenramblings.tagmo.eightbit.io.Debug.verbose
import com.hiddenramblings.tagmo.eightbit.io.Debug.warn
import com.hiddenramblings.tagmo.eightbit.io.Debug.isNewer
import com.hiddenramblings.tagmo.eightbit.io.Debug.isOlder
import com.hiddenramblings.tagmo.eightbit.os.Storage
import com.hiddenramblings.tagmo.nfctech.TagArray
import com.hiddenramblings.tagmo.security.SecurityHandler
import com.hiddenramblings.tagmo.widget.Toasty
import org.json.JSONException
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Paths
import java.text.ParseException
import java.util.regex.Pattern
import java.util.zip.ZipFile

class WebsiteFragment : Fragment() {
    private val webHandler = Handler(Looper.getMainLooper())
    private var mWebView: WebView? = null
    private var dialog: ProgressDialog? = null
    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.activity_webview, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mWebView = view.findViewById(R.id.webview_content)
        SecurityHandler(requireActivity(), object : SecurityHandler.ProviderInstallListener {
            override fun onProviderInstalled() {
                configureWebView()
            }

            override fun onProviderInstallException() {
                Toasty(requireActivity()).Long(R.string.fail_ssl_update)
            }

            override fun onProviderInstallFailed() {
                Toasty(requireActivity()).Long(R.string.fail_ssl_update)
            }
        })
    }

    @SuppressLint("AddJavascriptInterface", "SetJavaScriptEnabled")
    private fun configureWebView() {
        val webViewSettings = mWebView!!.settings
        mWebView!!.isScrollbarFadingEnabled = true
        webViewSettings.loadWithOverviewMode = true
        webViewSettings.useWideViewPort = true
        webViewSettings.allowFileAccess = true
        webViewSettings.allowContentAccess = false
        webViewSettings.javaScriptEnabled = true
        webViewSettings.domStorageEnabled = true
        webViewSettings.cacheMode = WebSettings.LOAD_NO_CACHE
        if (isOlder(Build.VERSION_CODES.KITKAT))
            @Suppress("DEPRECATION")
            webViewSettings.pluginState = WebSettings.PluginState.ON
        if (isNewer(Build.VERSION_CODES.LOLLIPOP)) {
            val assetLoader = WebViewAssetLoader.Builder().addPathHandler(
                "/assets/",
                AssetsPathHandler(requireContext())
            ).build()
            mWebView!!.webViewClient = object : WebViewClientCompat() {
                override fun shouldInterceptRequest(
                    view: WebView, request: WebResourceRequest
                ): WebResourceResponse? {
                    return assetLoader.shouldInterceptRequest(request.url)
                }

                override fun shouldOverrideUrlLoading(
                    view: WebView,
                    request: WebResourceRequest
                ): Boolean {
                    return if (request.url.lastPathSegment.equals("donate.html")) {
                        (requireActivity() as BrowserActivity).showDonationPanel()
                        true
                    } else {
                        super.shouldOverrideUrlLoading(view, request)
                    }
                }
            }
            if (WebViewFeature.isFeatureSupported(WebViewFeature.SERVICE_WORKER_BASIC_USAGE)) {
                ServiceWorkerControllerCompat.getInstance().setServiceWorkerClient(
                    object : ServiceWorkerClientCompat() {
                        override fun shouldInterceptRequest(
                            request: WebResourceRequest
                        ): WebResourceResponse? {
                            return assetLoader.shouldInterceptRequest(request.url)
                        }
                    })
            }
        }
        @Suppress("DEPRECATION")
        webViewSettings.allowFileAccessFromFileURLs =
            Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP
        @Suppress("DEPRECATION")
        webViewSettings.allowUniversalAccessFromFileURLs =
            Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP
        val download = JavaScriptInterface()
        mWebView!!.addJavascriptInterface(download, "Android")
        mWebView!!.setDownloadListener { url: String, _: String?, _: String?, mimeType: String, _: Long ->
            if (url.startsWith("blob") || url.startsWith("data")) {
                verbose(WebsiteFragment::class.java, url)
                mWebView!!.loadUrl(download.getBase64StringFromBlob(url, mimeType))
            }
        }
        loadWebsite(null)
    }

    fun loadWebsite(address: String?) {
        var website = address
        if (null != mWebView) {
            if (null == website) website = NFCIntent.SITE_GITLAB_README
            val webViewSettings = mWebView!!.settings
            if (website.startsWith(NFCIntent.SITE_GITLAB_README)) {
                webViewSettings.userAgentString = webViewSettings.userAgentString.replace(
                    ("(?i)" + Pattern.quote("android")).toRegex(), "TagMo"
                )
            }
            webViewSettings.setSupportZoom(true)
            webViewSettings.builtInZoomControls = true
            mWebView!!.loadUrl(website)
        } else {
            val delayedUrl = website
            webHandler.postDelayed({ loadWebsite(delayedUrl) }, TagMo.uiDelay.toLong())
        }
    }

    fun hasGoneBack() : Boolean {
        return if (mWebView!!.canGoBack()) {
            mWebView!!.goBack()
            true
        } else {
            false
        }
    }

    private inner class UnZip(var archive: File, var outputDir: File) :
        Runnable {
        @Throws(IOException::class)
        private fun decompress() {
            val zipIn = ZipFile(archive)
            val entries = zipIn.entries()
            while (entries.hasMoreElements()) {
                // get the zip entry
                val finalEntry = entries.nextElement()
                webHandler.post {
                    dialog!!.setMessage(getString(R.string.unzip_item, finalEntry.name))
                }
                if (finalEntry.isDirectory) {
                    val dir = File(
                        outputDir, finalEntry.name.replace("/", "")
                    )
                    if (!dir.exists() && !dir.mkdirs()) throw RuntimeException(
                        getString(R.string.mkdir_failed, dir.name)
                    )
                } else {
                    val `is` = zipIn.getInputStream(finalEntry)
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                        Files.copy(`is`, Paths.get(outputDir.absolutePath, finalEntry.name))
                    } else {
                        val fileOut = FileOutputStream(
                            File(outputDir, finalEntry.name)
                        )
                        val buffer = ByteArray(8192)
                        var len: Int
                        while (`is`.read(buffer).also { len = it } != -1) fileOut.write(
                            buffer,
                            0,
                            len
                        )
                        fileOut.close()
                    }
                    `is`.close()
                }
            }
            zipIn.close()
        }

        override fun run() {
            try {
                decompress()
            } catch (e: IOException) {
                warn(e)
            } finally {
                dialog!!.dismiss()
                archive.delete()
            }
        }
    }

    private fun saveBinFile(tagData: ByteArray, name: String) {
        try {
            val filePath = File(
                Storage.getDownloadDir(
                    "TagMo", "Downloads"
                ), "$name.bin"
            )
            val os = FileOutputStream(filePath, false)
            os.write(tagData)
            os.flush()
        } catch (e: IOException) {
            warn(e)
        }
    }

    private fun setBinName(base64File: String, mimeType: String) {
        val tagData =
            Base64.decode(base64File.replaceFirst("^data:$mimeType;base64,".toRegex(), ""), 0)
        val view = layoutInflater.inflate(R.layout.dialog_save_item, null)
        val dialog = AlertDialog.Builder(requireContext())
        val input = view.findViewById<EditText>(R.id.save_item_entry)
        try {
            val amiiboManager = getAmiiboManager(requireContext().applicationContext)
            input.setText(TagArray.decipherFilename(amiiboManager, tagData, true))
        } catch (e: IOException) {
            warn(e)
        } catch (e: JSONException) {
            warn(e)
        } catch (e: ParseException) {
            warn(e)
        }
        val backupDialog: Dialog = dialog.setView(view).create()
        view.findViewById<View>(R.id.button_save).setOnClickListener {
            saveBinFile(tagData, input.text.toString())
            backupDialog.dismiss()
        }
        view.findViewById<View>(R.id.button_cancel).setOnClickListener { backupDialog.dismiss() }
        backupDialog.show()
    }

    private inner class JavaScriptInterface {
        @JavascriptInterface
        @Throws(IOException::class)
        fun getBase64FromBlobData(base64Data: String) {
            convertBase64StringSave(base64Data)
        }

        fun getBase64StringFromBlob(blobUrl: String, mimeType: String): String {
            return if (blobUrl.startsWith("blob") || blobUrl.startsWith("data")) {
                "javascript: var xhr = new XMLHttpRequest();" +
                        "xhr.open('GET', '" + blobUrl + "', true);" +
                        "xhr.setRequestHeader('Content-type','" + mimeType + "');" +
                        "xhr.responseType = 'blob';" +
                        "xhr.onload = function(e) {" +
                        "  if (this.status == 200) {" +
                        "    var blobFile = this.response;" +
                        "    var reader = new FileReader();" +
                        "    reader.readAsDataURL(blobFile);" +
                        "    reader.onloadend = function() {" +
                        "      base64data = reader.result;" +
                        "      Android.getBase64FromBlobData(base64data);" +
                        "    }" +
                        "  }" +
                        "};" +
                        "xhr.send();"
            } else "javascript: console.log('Not a valid blob URL');"
        }

        @Throws(IOException::class)
        private fun convertBase64StringSave(base64File: String) {
            val zipType = getString(R.string.mimetype_zip)
            if (base64File.contains("data:$zipType;")) {
                val filePath = File(Storage.getDownloadDir("TagMo"), "download.zip")
                val os = FileOutputStream(filePath, false)
                os.write(
                    Base64.decode(
                        base64File.replaceFirst(
                            "^data:$zipType;base64,".toRegex(), ""
                        ), 0
                    )
                )
                os.flush()
                webHandler.post {
                    dialog = ProgressDialog.show(
                        requireContext(), "", "", true
                    )
                }
                Thread(UnZip(
                        filePath, Storage.getDownloadDir("TagMo", "Downloads")
                )).start()
            } else {
                val binTypes = resources.getStringArray(R.array.mimetype_bin)
                for (binType in binTypes) {
                    if (base64File.contains("data:$binType;")) {
                        setBinName(base64File, binType)
                        break
                    }
                }
            }
        }
    }
}